lib/streamlit/runtime/state/session_state.py

```diff
@@ -488,8 +488,8 @@ def on_script_will_rerun(self, latest_widget_states: WidgetStatesProto) -> None:
         Update widget data and call callbacks on widgets whose value changed
         between the previous and current script runs.
         """
-        # Update ourselves with the new widget_states. The old widget states,
-        # used to skip callbacks if values haven't changed, are also preserved.
+        # Clear any triggers that weren't reset because the script was disconnected
+        self._reset_triggers()
```

Comment:
As an alternative, I considered splitting `on_script_finished` in two, and calling the part with the trigger reset from `request_stop`, before the disconnect. I opted to do it here instead to be more sure that the reset will happen, and to avoid having to create a bunch of plumbing methods that now correspond to calling just one function on the actual session state.